### PR TITLE
refactor: rename to OpenTelemetry.Extensions.Hosting from OpenTelemetry.Hosting

### PR DESCRIPTION
**Pull request by longility**
In terms of usage, consumers of this package has a primary intent to hook in using DependencyInjection and nothing to do with Hosting. Hosting is unfortunately needed for OpenTelemetry and is an implementation detail that the consumer of this package does not need to know. By renaming to `OpenTelemetry.Extensions.DependencyInjection`, this allows consumer to intuitively select this package, knowing that this is what is needed to hook in OpenTelemetry.

There is also a minor dependency change to `Microsoft.Extensions.Hosting.Abstractions` from `Microsoft.Extensions.Hosting` for this csproj.

### CODE REVIEW

**Comment by longility** [comment]
This could be controversial, but I wanted to take a stab to see what others think...while I'm down the rabbit hole of refactoring.

**Comment by cijothomas** [comment]
@longility Thanks. I agree Hosting is bit confusing and not truly explaining the intent,

Would OT.ConfigurationExtensions, or simply OT.Extensions be better? I dont have strong opinion on this or your proposal.

**Comment by CodeBlanch** [comment]
One thought I have had in this area...

There's a lot of packages under the OT umbrella. To get up and running in an aspnetcore project most people are probably going to need OpenTelemetry.Collector.AspNetCore, OpenTelemetry.Collector.Dependencies, OpenTelemetry.Hosting, and then their exporter of choice.

It might be nice to have a meta package that combines OpenTelemetry.Collector.AspNetCore, OpenTelemetry.Collector.Dependencies, & OpenTelemetry.Hosting into something like OpenTelemetry.AspNetCore? Basically so you only have to consume 2: the meta & your exporter. I think it would help resolve the naming weirdness because people wouldn't need to care, just grab the meta package and off you go.

How's that for controversy :)

**Comment by cijothomas** [comment]
> One thought I have had in this area...
> 
> There's a lot of packages under the OT umbrella. To get up and running in an aspnetcore project most people are probably going to need OpenTelemetry.Collector.AspNetCore, OpenTelemetry.Collector.Dependencies, OpenTelemetry.Hosting, and then their exporter of choice.
> 
> It might be nice to have a meta package that combines OpenTelemetry.Collector.AspNetCore, OpenTelemetry.Collector.Dependencies, & OpenTelemetry.Hosting into something like OpenTelemetry.AspNetCore? Basically so you only have to consume 2: the meta & your exporter. I think it would help resolve the naming weirdness because people wouldn't need to care, just grab the meta package and off you go.
> 
> How's that for controversy :)

This is very good. For the product I work (Microsoft Application Insights), we created Microsoft.ApplicationInsights.AspNetCore for Asp.Net Core projects, 
Microsoft.ApplicationInsights.WorkerService for non web projects (worker, background etc)
Microsoft.ApplicationInsights.Web for Asp.Net projects. These are essentially meta packages for each 'type' of application. (exporter is implicit for application insights world, for OT - user picks the metapackage for their type of app, and exporter - this sounds very good!)

**Comment by longility** [comment]
Funny, on my side project I have a meta package as well. New proposal on work to be done based on above proposal:
1. empty project `OpenTelemetry.AspNetCore`
  - depends on `OpenTelemetry.Dependencies`
  - depends on `OpenTelemetry.Adapter.AspNetCore`
2. documentation of approach
Can we go ahead with above proposal or should we wait for certain people to approve?

I think this PR change is still a good refactor, but it is less important once we have the proposed meta-package as mentioned by @CodeBlanch and how people shouldn't care anymore. If I don't get much feedback on taking this on, then I will go ahead and close soon.

**Comment by longility** [comment]
Just to give an idea on how I decided the name is based on searching on nuget.org https://www.nuget.org/packages?q=Extensions.DependencyInjection

MassTransit, Mediatr/Automapper, and AppMetrics follows what I put in the PR. I haven't seen other patterns out there that makes me lean elsewhere.

**Comment by cijothomas** [comment]
@longility Can you create a new issue describing the Meta package approach, so we can get feedback from others as well?

**Comment by longility** [comment]
New issue created for meta package proposal.

Keeping this PR open as I think this is still a nice little improvement. Seeing if others agree if it is a worthwhile change to add in or not.

**Comment by cijothomas** [comment]
Just additional note:
@longility Though the name OT.Extensions.DI sound better than Hosting, the package is not taking dependency on MS.Extensions.DI, instead it depends on MS.Extens.Hosting.Abstractions. (Hosting is needed to kick off the `TelemetryFactoryHostedService`.

**Comment by longility** [comment]
Yep I noticed that when I tried to remove it to see what happens.

However, I think from library consumer perspective, I would think we should name it in a way that makes it more intuitive on what the library is for or what library I would want. I think for a library consumer perspective, I'm looking to hook into DI. I really don't care that there is a hosted service.

**Comment by CodeBlanch** [comment]
If I had to vote it would probably be for OpenTelemetry.Extensions.Hosting for the reasons @cijothomas pointed out. "Hosting" to me is a superset that includes "DependencyInjection" and other things you need to do to build up an application. For me it's indicative that if you are using `Microsoft.Extensions.Hosting.IHostBuilder` than this library is for you. If you are using `Microsoft.Extensions.DependencyInjection.IServiceCollection` by itself with some kind of custom host, then this library really won't get you up and running. Because you'll have to manage the lifetime of the `TracerFactory` on your own (that's what the `IHostedService` does). You will probably also not want the hosting dependency in the later case.

**Comment by longility** [comment]
Interesting points @CodeBlanch . I think `OpenTelemetry.Extensions.Hosting` is an improvement to `OpenTelemetry.Hosting` so I can change that when I have time.

**Comment by longility** [comment]
@cijothomas @CodeBlanch this is now updated based on feedback and ready to be reviewed/merged.

File: OpenTelemetry.sln [hunk 1]
```diff
@@ -43,7 +43,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenTelemetry.Exporter.Appl
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenTelemetry.Exporter.Stackdriver", "src\OpenTelemetry.Exporter.Stackdriver\OpenTelemetry.Exporter.Stackdriver.csproj", "{DE1B4783-C01F-4672-A6EB-695F1717105B}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenTelemetry.Api", "src\OpenTelemetry.Api\OpenTelemetry.Api.csproj", "{99F8A331-05E9-45A5-89BA-4C54E825E5B2}"
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "OpenTelemetry.Api", "src\OpenTelemetry.Api\OpenTelemetry.Api.csproj", "{99F8A331-05E9-45A5-89BA-4C54E825E5B2}"
```
Comment (cijothomas) [inline OpenTelemetry.sln:46]: change to the API project is not part of this PR right? wondering why this has changed

Comment (longility) [reply OpenTelemetry.sln:46]: Yes, that should not be changed...let me fix that. Sorry for the noise.
